nixos/immich: init at 1.79.1

oddlama opened the pull request:
###### Description of changes

This is an early draft to add a package and module for Immich, a very popular self-hosted photo and video backup solution. I would appreciate any feedback on module ergonomics. So far, all important settings are already exposed in the module and the packages should build correctly.

- Depends on #244233
- Depends on #244655 (currently includes a workaround that can be removed once this is merged)

TODOs:

- [ ] The nixos test would require network access and could benefit from persisting the machine learning models if that is possible (otherwise this is a multi-GB download each time the test is run)
- [ ] Uploading images / videos could invoke external tools such as ffmpeg, exiftool, ..., and at this point I haven't tested if all required tools are propagated.
- [x] An update script is required to also update the sub-packages and their npmDepHashes.
- [ ] aarch64 support should be possible and should be tested.

###### Things done

**Early draft, not everything can be tested yet.**

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

AndersonTorres commented:
Split the commit
- one for the package
- one for nixos module
- one for module tests

oddlama committed:
immich: init at 1.79.1

oddlama committed:
nixos/immich: init module

oddlama committed:
nixos/tests/immich: init tests

oddlama commented:
updated to 1.70.0, added automatic update script

qbit commented:
It seems `nodePackages.immich` is a thing, not sure if the `-cli` bit should be changed to use that or not?

oddlama commented:
> It seems `nodePackages.immich` is a thing, not sure if the `-cli` bit should be changed to use that or not?

That's the same thing that can be built from source here, just downloaded from the package registry. There currently are several cli implementations, and the one packaged here is the endorsed CLI that will replace all the others in the future.

mattmelling commented on nixos/modules/services/web-apps/immich.nix:
```diff
@@ -0,0 +1,574 @@
+      else "export DB_PASSWORD=$(cat ${serverCfg.postgres.passwordFile})"
+    )}
+
+    ${optionalString serverCfg.typesense.enable ''
+      export TYPESENSE_API_KEY=$(cat ${serverCfg.typesense.apiKeyFile})
+    ''}
```
immich-server looks for `TYPESENSE_API_KEY` even if `TYPESENSE_ENABLE=false`, so when `cfg.server.typesense.enable = false`, immich-server won't start.

mattmelling commented:
I can get as far as the admin user sign up screen, hitting sign up causes an xhr to fail with a 500

Atemu commented:
This needs a rebase as it does not build with the newer versions of its dependencies we have in master right now.

dotlambda commented on pkgs/servers/immich/default.nix:
```diff
@@ -0,0 +1,238 @@
+    inherit (sources) hash;
+  };
+
+  cli = buildNpmPackage' {
```
There is already a package `immich-cli`.

oddlama replied:
Yes but here we have the original source and the existing immich-cli just pulls it from npm. Either they should live on as separate packages like they do now, or the immich-cli should be removed.

dotlambda replied:
No it was repackaged to use `buildNpmPackage` in https://github.com/NixOS/nixpkgs/pull/250280.

dotlambda replied:
Upstream should add a remark about that to the former repo's readme. Maybe you wanna make a PR?

dotlambda commented on pkgs/servers/immich/default.nix:
```diff
@@ -0,0 +1,238 @@
+  cli = buildNpmPackage' {
+    pname = "immich-cli";
+    inherit version;
+    src = "${src}/cli";
```
Do you know what the difference from https://github.com/immich-app/CLI is?

oddlama replied:
See comment above

Atemu commented on nixos/modules/services/web-apps/immich.nix:
```diff
@@ -0,0 +1,575 @@
+          description = "Whether to open the firewall for the specified port.";
+        };
+
+        extraConfig = mkOption {
```
```suggestion
        settings = mkOption {
```

https://github.com/NixOS/rfcs/pull/42

Atemu commented on nixos/modules/services/web-apps/immich.nix:
```diff
@@ -0,0 +1,575 @@
+
+    systemd.services.immich-machine-learning = mkIf mlCfg.enable {
+      description = "Immich machine learning (Self-hosted photo and video backup solution)";
+      after = ["network.target"];
```
Nit: Please add spaces around brackets:

```suggestion
      after = [ "network.target" ];
```

Atemu reviewed (changes requested):
* [Immich requires a reverse proxy](https://github.com/immich-app/immich/blob/41461e0d5dde9675d462af8bbe30c58975dd754e/docs/docs/administration/reverse-proxy.md#reverse-proxy). The module needs to handle that. This prevents the webapp from doing pretty much anything other than displaying the registration screen because the API is unavailable (reverse proxy needs to redirect it to `IMMICH_SERVER_URL`).
* The ML package is broken because our onnx still requires protobuf3. It can be built against pythonPackages.protobuf 4 but only when that package is supplied with `protobuf3_21` which uses an older version of `abseil-cpp`.  
  Additionally, it now requires open-clip-torch which seems to have a hard dependency on protobuf3 :/  
  At this point, I'd just mark the ML package as broken and disable the service by default. It's a very desirable feature but not core to the app, so that would be fine IMO.
* The user is required to set up postgres and typesense themselves. I don't know how this is handled in other modules but I think this should at least be mentioned in the `.enable` option but perhaps also offer an optional sane default config.

Atemu commented:
With a little nginx config like this

```Nix
    {
      locations = {
        "/" = {
          proxyPass = "http://127.0.0.1:${toString webCfg.port}";
        };
        "${webCfg.apiUrlExternal}/" = {
          proxyPass = "${webCfg.serverUrl}/";
        };
      };
    };
```

and a bit of faffing about with db and search gets us a working Immich on NixOS:

![image](https://github.com/NixOS/nixpkgs/assets/18599032/63cef7d8-2d9e-4c0b-9551-2822a7f08439)

teto commented:
failed to build on unstable because of python insight broken package, willing to test though

dotlambda commented:
> failed to build on unstable because of python insight broken package

Should be solved by using
```nix
python = python3.override {
  packageOverrides = self: super: {
    protobuf = super.protobuf.override {
      inherit (pkgs.onnxruntime) protobuf;
    };
  };
};
```

teto commented:
I dont think it does
```
       … while evaluating derivation 'immich-machine-learning-1.75.1'
         whose name attribute is located at /nix/store/mj9drg7669lnca25lmrssjnslanxr5fs-source/pkgs/stdenv/generic/make-derivation.nix:300:7

       … while evaluating attribute 'propagatedBuildInputs' of derivation 'immich-machine-learning-1.75.1'

         at /nix/store/mj9drg7669lnca25lmrssjnslanxr5fs-source/pkgs/stdenv/generic/make-derivation.nix:354:7:

          353|       depsHostHostPropagated      = lib.elemAt (lib.elemAt propagatedDependencies 1) 0;
          354|       propagatedBuildInputs       = lib.elemAt (lib.elemAt propagatedDependencies 1) 1;
             |       ^
          355|       depsTargetTargetPropagated  = lib.elemAt (lib.elemAt propagatedDependencies 2) 0;

       … from call site

         at /nix/store/mj9drg7669lnca25lmrssjnslanxr5fs-source/pkgs/stdenv/generic/check-meta.nix:450:16:

          449|         {
          450|           no = handleEvalIssue { inherit meta attrs; } { inherit (validity) reason errormsg; };
             |                ^
          451|           warn = handleEvalWarning { inherit meta attrs; } { inherit (validity) reason errormsg; };

       … while calling 'handleEvalIssue'

         at /nix/store/mj9drg7669lnca25lmrssjnslanxr5fs-source/pkgs/stdenv/generic/check-meta.nix:226:38:

          225|
          226|   handleEvalIssue = { meta, attrs }: { reason , errormsg ? "" }:
             |                                      ^
          227|     let

       error: Package ‘python3.11-insightface-0.7.3’ in /nix/store/mj9drg7669lnca25lmrssjnslanxr5fs-source/pkgs/development/python-modules/insightface/default.nix:75 is marked as broken, refusing to evaluate.

       a) To temporarily allow broken packages, you can use an environment variable
          for a single invocation of the nix tools.

            $ export NIXPKGS_ALLOW_BROKEN=1

        Note: For `nix shell`, `nix build`, `nix develop` or any other Nix 2.4+
        (Flake) command, `--impure` must be passed in order to read this
        environment variable.

       b) For `nixos-rebuild` you can set
         { nixpkgs.config.allowBroken = true; }
       in configuration.nix to override this.
```

dotlambda commented:
You have to remove `meta.broken` from insightface of course.

oddlama commented:
> @oddlama Are you still working on this? I'm interested in using thins on Nixos

Yes of course, I'll keep update this from time to time.
Due to the protobuf 3->4 complications I wanted to let the ecosystem settle a bit before continuing. I will push an update in the next days.

I'm mostly requiring feedback on the module ergonomics at this stage. To get this out of draft stage though we'll have to wait a bit longer since there will be some significant changes to immich (possibly typesense -> pgvector) that I would like to await before pulling this into master.

dotlambda commented:
> Due to the protobuf 3->4 complications I wanted to let the ecosystem settle a bit before continuing. I will push an update in the next days.

Isn't that solved by my comment?

oddlama commented:
> Isn't that solved by my comment?

Possibly, I haven't tested it yet. Did you run all tests of all dependencies to see whether any of these fails due to the different protobuf version? protobuf is known to cause issues when changing major versions, so I'm not immediately sure that all dependencies support it.

dotlambda commented:
> Did you run all tests of all dependencies to see whether any of these fails due to the different protobuf version?

No.

> protobuf is known to cause issues when changing major versions, so I'm not immediately sure that all dependencies support it.

That's why the solution is going back to the older version.

oddlama commented:
> [!WARNING]
> I've tried updating this to v1.79.1, but ran into issues with the server's updated sharp dependency, which now tries to download stuff at install time which is prohibited by the sandbox. Since this is done in the dependency download step in `buildNpmPackage`, there is no easy workaround at this time. (Or at least none that I am aware of).
>
> I'll have to revisit this later. If you know a workaround, please let me know.
>
> ```
> npm ERR! code 1
> npm ERR! path /build/server/node_modules/sharp
> npm ERR! command failed
> npm ERR! command sh -c (node install/libvips && node install/dll-copy && prebuild-install) || (node install/can-compile && node-gyp rebuild && node install/dll-copy)
> npm ERR! sharp: Downloading https://github.com/lovell/sharp-libvips/releases/download/v8.14.5/libvips-8.14.5-linux-x64.tar.br
> npm ERR! sharp: Please see https://sharp.pixelplumbing.com/install for required dependencies
> npm ERR! sharp: Installation error: getaddrinfo EAI_AGAIN github.com
> 
> npm ERR! A complete log of this run can be found in: /build/cache/_logs/2023-10-01T00_51_29_351Z-debug-0.log

- Rebased on nixpkgs/master and updated to immich v1.79.1
- Added packages for new required dependencies `clip-server` and `open-clip-torch`
- All packages except the server are building, but beware that the nixos tests could not run yet.

If you want to test this PR right now, please refer to the [previous version](https://github.com/NixOS/nixpkgs/commit/ce90a106a1aa939a2bc73ce12cbeb35ead0cc87e).

christoph-heiss commented:
> Warning I've tried updating this to v1.79.1, but ran into issues with the server's updated sharp dependency, which now tries to download stuff at install time which is prohibited by the sandbox. Since this is done in the dependency download step in `buildNpmPackage`, there is no easy workaround at this time. (Or at least none that I am aware of).
> 
> I'll have to revisit this later. If you know a workaround, please let me know.

One thing that can be done is to pin (by overriding) the version of libvips to **exactly** the version it wants, something like this should work: 
```nix
libvips = pkgs.vips.overrideAttrs (_: rec {
  version = "8.14.5";
  src = pkgs.fetchFromGitHub {
    owner = "libvips";
    repo = "libvips";
    rev = "v${version}";
    hash = "sha256-TzuXYKzTSptYb1fLOFzfsvsN4NWsmNfB6LUPVpaeEAg=";
    # Remove unicode file names which leads to different checksums on HFS+
    # vs. other filesystems because of unicode normalisation.
    postFetch = ''
      rm -r $out/test/test-suite/images/
    '';
  };
});
```
IMHO completely insane that libvips tries to download something during build time.

Scrumplex commented:
The recent release of immich 1.88.x merged the -web component into the backend service. So this module can probably be simplified now.

See https://github.com/immich-app/immich/discussions/5086

pinpox commented:
> The recent release of immich 1.88.x merged the -web component into the backend service. So this module can probably be simplified now.

@oddlama I had a look at updating your module to incorporate this, but didn't manage to make it work. It seems the `immich-web` systemd service can be fully removed and it's config merged with the `immich-server` config. If you get a version running, I can offer help with testing.

Scrumplex commented:
Yet another upcoming upstream change! Typesense will be dropped in favor of Postgres with the pgvecto.rs extension. As far as I can tell, the latter is not packaged yet

See https://github.com/immich-app/immich/releases/tag/v1.90.2

Atemu commented:
I'm slowly starting to question whether we should bother packaging at the speed at which upstream moves...

Scrumplex commented:
> I'm slowly starting to question whether we should bother packaging at the speed at which upstream moves...

I am personally very interested in this, as I don't really like my current Docker-Compose and previously Helm deployments of Immich.

At least upstream changes mostly reduce complexity, although the Postgres extension does feel a bit more annoying to implement than having a separate daemon

oddlama commented:
Sorry for the slow update on this, I'm unfortunately quite busy at this time of year :/

> I'm slowly starting to question whether we should bother packaging at the speed at which upstream moves...

I agree, this is becoming a concern more than I initially assumed. I've been in close contact with the immich devs for some time, so I've been aware of these big changes already, but they are still pushing out new stuff so insanely fast that it is hard to keep up. 

Their architecture is just very much in motion right now and their development process (while very effective) is just docker-centric. That means there are quite some things in immich which assume exactly those conditions and that makes the module development a bit rough sometimes 😅. The current lack of E2E tests (which they are working on) means that I'll only discover these cases at runtime.

What concerns me most regarding this module is that each time they change something about their architecture (which is often) it will likely cause downstream breaking changes for users of this module.

> I am personally very interested in this, as I don't really like my current Docker-Compose and previously Helm deployments of Immich.

Me too (obviously :P). But as long as native deployments are not a core goal of immich this will mean a lot of work and breakages on the NixOS side - and if I understand correctly that won't be the case for a long time, if ever.

Maybe the way to go forward here is to convert this draft to its own `immich-nixos` project that we can update more frequently and easily by ad-hoc packaging what is needed. That way we don't have to wait until all dependencies are available in NixOS proper. If anyone has their own thoughts on this matter, please share them :)

I'll definitely try to get all the changes in here sometime, but it'll take some work.

christoph-heiss commented:
> Maybe the way to go forward here is to convert this draft to its own `immich-nixos` project that we can update more frequently and easily by ad-hoc packaging what is needed. That way we don't have to wait until all dependencies are available in NixOS proper. If anyone has their own thoughts on this matter, please share them :)

That definitely sounds like a idea to pursue. I'd also would very much like to deploy/use Immich and be ready to help with such an effort.

You already listed a bunch of good reason for this, and having the possibility to move fast (just like upstream) would make things a lot easier IMHO. In addition to that, it'd enable much easier collaboration on the effort to keep up with upstream and provide a working version.

If, in the future, the rapid development with breakages all the time slows down a bit, it can still be integrated in nixpkgs after all.

Just my 2ct.

diogotcorreia commented:
For those running Immich inside Docker but with PostgreSQL under NixOS, to update to 1.91.0 you need the `pgvecto.rs` PostgreSQL extension.

I've managed to (kinda) package it with the release binary (I did not try to build from source, seems to be a bit complicated because you need `cargo-pgrx`).
Here is the commit in my config updating to 1.91.0 if you want to copy the package: https://github.com/diogotcorreia/dotfiles/commit/7676201683a3785ef17eff9f4ad3295375c670bd
Suggestions are welcome!

bhankas commented:
Since there seems to be an agreement on separate project being better suited for now, should we ask for a repo under nix-community? Ideally that would be better, but if we don't want that to be a blocker, we can always get started with any of the contributor's github and then transfer ownership. What say?

Patryk27 commented:
> should we ask for a repo under nix-community

I think having flake.nix inside [the official repository](https://github.com/immich-app/immich) would be more useful, if upstream is willing.

bhankas commented:
Since upstream devs don't use Nix for development (otherwise they'd already have a *.nix file), it would be undue burden on them to either keep the flake updated, or keep reviewing+accepting PRs for it, or provide someone full commit rights. I believe that is not a reasonable ask.

Besides, the idea behind separate repo for nixospackage+module is to be detached from both up streams (immich and nixpkgs) so we can keep working on the module itself.

Combined with f-droid allowing exact version of app to be and remain installed, both client and server can stay in sync until the package+module gets updated for upstream immich changes.

adamcstephens commented:
> Since there seems to be an agreement on separate project being better suited for now, should we ask for a repo under nix-community? Ideally that would be better, but if we don't want that to be a blocker, we can always get started with any of the contributor's github and then transfer ownership. What say?

Just create a repo, start building and invite maintainers if you want. No need for it to start under nix-community.

pinpox commented:
> > Since there seems to be an agreement on separate project being better suited for now, should we ask for a repo under nix-community? Ideally that would be better, but if we don't want that to be a blocker, we can always get started with any of the contributor's github and then transfer ownership. What say?
> 
> Just create a repo, start building and invite maintainers if you want. No need for it to start under nix-community.

Feel free to ping or invite me, I'd like to test this.

GetPsyched commented:
Have there been any updates on this effort?
cc @oddlama

oddlama commented:
No sorry, currently I don't have this updated. Immich has been moving extremely fast recently, and they often introduce new dependencies which (especially with python ML stuff) requires a lot of work to package. I haven't had the time to keep it up with their pace.

If anyone wants to pick this up, feel free to do so. But be aware that this won't be a walk in the park to keep updated :P
Maybe I'll revisit this when immich goes stable.

anoadragon453 commented:
Here is my NixOS Immich setup in case it is useful to anyone here. I am currently running Immich v1.106.4.

https://github.com/anoadragon453/dotfiles/blob/main/modules/server/immich.nix

Note that to get pgvecto-rs to build (a postgres extension that Immich relies on) using current `nixos-unstable`, https://github.com/NixOS/nixpkgs/pull/320302 is needed.

RomeoV commented:
Thanks @anoadragon453, this looks like it works for me. For others struggling to get it to work:
1. Copy Andrew's dotfile into e.g. `/etc/nixos/immich.nix`. If you typically use a stable release, change `pkgs` to `pkgs_unstable` and add a line
```nix
{config, lib, pkgs_unstable, ...}:
let
  pkgs = pkgs_unstable;
in 
...
```
3. Add something like this to your `configuration.nix`
```nix
{ config, lib, pkgs, ... }: 
let
  unstable = import <nixos-unstable> {};
  pkgs_unstable = unstable.pkgs;
in {
  imports =   [
    (import /etc/nixos/immich.nix {inherit lib config pkgs_unstable; })
  ];

  services.immich = {
    enable = true;
    domain = "immich.your-domain.xyz";
    port = 8006;
    metricsPortServer = 8009;
    metricsPortMicroservices = 8010;
    storagePath = "/mnt/storage-box/immich";
    logLevel = "log";
  };
```
4. et voilà

jvanbruegge commented:
If someone does not want to run docker containers for immich and would like a native nixos package instead, I've packaged the current version and also built a nixos module: https://github.com/jvanbruegge/server-config/tree/master/pkgs/immich

If you use flakes, you should be able to include the module like this:
```nix
{
  description = "NixOS configuration";

  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs/nixos-unstable";
    immich.url = "github:jvanbruegge/server-config";
    immich.inputs.nixpkgs.follows = "nixpkgs";
  };

  outputs = inputs@{ nixpkgs, immich, ... }: {
    nixosConfigurations = {
      hostname = nixpkgs.lib.nixosSystem {
        system = "x86_64-linux";
        modules = [
          ./configuration.nix
          immich.nixosModules.immich
        ];
      };
    };
  };
};
```
The flake also exports just the package (without the module). I have this running for a few weeks now on my own server

marcusramberg commented:
> The flake also exports just the package (without the module). I have this running for a few weeks now on my own server.

https://github.com/jvanbruegge/server-config/blob/5ee96717d8476c6671bb1c28d144007bdf74915b/flake.nix#L41 seems like your flake currently does not export the module at all?

jvanbruegge commented:
> seems like your flake currently does not export the module at all?

Sorry about that, that got lost in rebasing. Fixed now.

> Would you be willing to create a PR for the package and optionally the module too?

Sure, I can do so soon

christoph-heiss commented:
@Pablito2020  

> ```
> $ chmod -R 0777 /data
> ```

Please **never** do that! Are you aware of the consequences?
It should be owned by the `immich` user & group only, without allowing every users accessing it.

Pablito2020 commented:
> @Pablito2020
> 
> > ```
> > $ chmod -R 0777 /data
> > ```
> 
> Please **never** do that! Are you aware of the consequences? It should be owned by the `immich` user & group only, without allowing every users accessing it.

Yes, I am aware of the consequences of giving rwx access to all users of the machine, it was just for getting it to work, someone should add the privileges to the systemd-service inside the module. 

I was just trying to tell what's missing inside the module for making it work on the first try.

I am sorry if my comment wasn't clear.

christoph-heiss commented:
Derailed into OT now completely, but:

> Yes, I am aware of the consequences of giving rwx access to all users of the machine, it was just for getting it to work, someone should add the privileges to the systemd-service inside the module.

It is just that people _really_ like to blindly run random commands people post on the internet without knowing what they do, inevitable resulting in losing data or such in the end ..

Sometimes you (unfortunately!) have to protect people from themselves.

jvanbruegge commented:
I've now opened the PR for the package and module here: https://github.com/NixOS/nixpkgs/pull/324127

Differences to the version in my repo:
- The cli executable is just called `immich` as that is what the immich docs use
- All configuration options in the module have documentation now
- The postgres config by default uses the unix socket now, meaning that `services.immich.enable = true` is enough to get a working service.
- I've added a nixos test that creates and admin user, creates an API key for that admin user and then uses the immich-cli to get the server-info (so that both server and cli get tested)
